File: 2024/TypeScript/1/solution.ts
import { readInputFile } from "../../../lib.js";

function solution(data: string): void {
    const list1: number[] = [];
    const list2: number[] = [];

    data.split("\r\n").forEach((line) => {
        line = line.trim();
        if (line) {
            const parts = line.split(/\s+/);

            list1.push(Number(parts[0]));
            list2.push(Number(parts[1]));
        }
    });

    const list1Sorted = list1.sort((a, b) => a - b);
    const list2Sorted = list2.sort((a, b) => a - b);

    console.log("list1 sorted:", list1Sorted);
    console.log("list2 sorted:", list2Sorted);

    const pairs = list1Sorted.map((item, index) => [item, list2Sorted[index]]);

    console.log("pairs:", pairs);

    const distancePerPair = pairs.map(([a, b]) => Math.abs(a - b));

    const totalDistance = distancePerPair.reduce(
        (acc, distance) => acc + distance,
        0
    );

    console.log("total distance between all pairs:", totalDistance);

    const totalSimilarityScore = list1Sorted
        .map((value) => {
            const amountOfOccurrencesInList2 = list2Sorted.filter(
                (i) => i === value
            ).length;

            return value * amountOfOccurrencesInList2;
        })
        .reduce((acc, score) => acc + score, 0);

    console.log("total similarity score of the 2 lists:", totalSimilarityScore);
}

readInputFile(solution);

File: 2024/C++/1/solution.cpp
#include <fstream>
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

void main()  {
    ifstream input_stream("input.txt");

    if (!input_stream) cerr << "Can't open input file!";

    vector<string> text;
    string line;

    while (getline(input_stream, line)) {
        text.push_back(line);
    }

    vector<int> list1;
    vector<int> list2;

    for (string line : text) {
        int item1 = stoi(line.substr(0, 5));
        int item2 = stoi(line.substr(8, 5));

        list1.push_back(item1);
        list2.push_back(item2);
    }

    sort(list1.begin(), list1.end());
    sort(list2.begin(), list2.end());

    int totalDistance = 0;
    int totalSimilarityScore = 0;

    for (int i = 0; i < list1.size(); i++) {
        totalDistance += abs(list1[i] - list2[i]);

        int item1 = list1[i];

        int amountOfOccurrencesInList2 = count(list2.begin(), list2.end(), item1);

        totalSimilarityScore += item1 * amountOfOccurrencesInList2;
    }

    cout << "Total distance is: " << totalDistance << endl;
    cout << "Total similarity score is: " << totalSimilarityScore << endl;
}